crates/ark/src/lsp/diagnostics_syntactic.rs

```diff
@@ -0,0 +1,559 @@
+fn diagnose_error(
+    node: Node,
+    context: &DiagnosticContext,
+    diagnostics: &mut Vec<Diagnostic>,
+) -> anyhow::Result<()> {
+    let mut report = node.is_error();
+    let mut cursor = node.walk();
+
+    for child in node.children(&mut cursor) {
+        if node_has_error(&child) {
+            // At least one child is also an `ERROR` node, so we
+            // definitely won't report ourselves as an `ERROR` anymore.
+            report = false;
+
+            diagnose_error(child, context, diagnostics)?;
+        }
+    }
+
+    if report {
+        diagnostics.push(build_syntax_diagnostic(node, context)?);
+    }
+
+    Ok(())
+}
+
+fn build_syntax_diagnostic(node: Node, context: &DiagnosticContext) -> anyhow::Result<Diagnostic> {
+    if let Some(diagnostic) = build_syntax_diagnostic_missing_open(node, context)? {
+        return Ok(diagnostic);
+    }
+
+    Ok(build_syntax_diagnostic_default(node, context))
+}
+
+// Use a heuristic that if we see a syntax error and it just contains a `)`, `}`, or `]`,
+// then it is probably a case of missing a matching open token.
+fn build_syntax_diagnostic_missing_open(
+    node: Node,
+    context: &DiagnosticContext,
+) -> anyhow::Result<Option<Diagnostic>> {
+    let text = context.contents.node_slice(&node)?;
+
+    let open_token = if text == ")" {
+        "("
+    } else if text == "}" {
+        "{"
+    } else if text == "]" {
+        "["
+    } else {
+        // Not an unmatched closing token
+        return Ok(None);
+    };
+
+    let range = node.range();
+
+    Ok(Some(new_missing_open_diagnostic(
+        open_token, range, context,
+    )))
+}
+
+fn build_syntax_diagnostic_default(node: Node, context: &DiagnosticContext) -> Diagnostic {
+    let range = node.range();
+    let row_span = range.end_point.row - range.start_point.row;
+
+    if row_span >= 20 {
+        return build_syntax_diagnostic_truncated_default(range, context);
+    }
+
+    // The most common case, a localized syntax error that doesn't span too many rows
+    let message = String::from("Syntax error");
+    new_syntax_diagnostic(message, range, &context)
+}
+
+// If the syntax error spans more than 20 rows, just target the starting position
+// to avoid overwhelming the user.
+fn build_syntax_diagnostic_truncated_default(
+    range: Range,
+    context: &DiagnosticContext,
+) -> Diagnostic {
+    // In theory this is an empty range, as they are constructed like `[ )`, but it
+    // seems to work for the purpose of diagnostics, and getting the correct
+    // coordinates exactly right seems challenging.
+    let start_range = Range {
+        start_byte: range.start_byte,
+        start_point: range.start_point,
+        end_byte: range.start_byte,
+        end_point: range.start_point,
+    };
+
+    // `+1` because it is user facing and editor UI is 1-indexed
+    let end_row = range.end_point.row + 1;
+    let message = format!("Syntax error. Starts here and ends on row {end_row}.");
```

Comment:
I think that's more telling for users:

```suggestion
    let message = format!("Syntax error. Starts here and ends on line {end_row}.");
```